src/dparse/ast.d

```diff
@@ -3954,3 +3959,75 @@ unittest // Support GCC-sytle asm statements
     });
 }
 
+unittest // Support throw expressions nested in other expressions
+{
+    import dparse.lexer, dparse.parser, dparse.rollback_allocator;
+
+    static void testThrow(const string source, void function(ThrowExpression, UnaryExpression throwable) cb)
+    {
+        static final class ThrowChecker : ASTVisitor
+        {
+            alias visit = ASTVisitor.visit;
+
+            ThrowExpression throwExp;
+
+            override void visit(const ThrowExpression throwExp)
+            {
+                this.throwExp = cast() throwExp;
+            }
+        }
+
+        RollbackAllocator ra;
+        LexerConfig cf = LexerConfig("", StringBehavior.source);
+        StringCache ca = StringCache(16);
+
+        auto tokens = getTokensForParser(source, cf, &ca);
+        Module m = parseModule(tokens, "", &ra);
+
+        scope check = new ThrowChecker();
+        check.visit(m);
+
+        assert(check.throwExp);
+        auto arg = check.throwExp.throwable;
+        auto unaExp = cast(UnaryExpression) arg;
+        assert(unaExp, arg ? arg.toString() : "null");
+
+        cb(check.throwExp, unaExp);
+    }
+
+    // ThrowExpression may appear nested in other statements / expressions
+    testThrow(q{
+        int main()
+        {
+            return foo ? bar : throw new Exception("Hello, World!");
+        }
+    },
+    (throwExp, throwable)
+    {
+        auto type = throwable.newExpression.type;
+        auto id = type.type2.typeIdentifierPart.identifierOrTemplateInstance.identifier;
+        assert(id.text == "Exception");
+
+        const item = throwable.newExpression.arguments.argumentList.items[0];
+        assert(item);
+        const una = cast(UnaryExpression) item;
+        assert(una, typeid(item).toString());
+        assert(una.primaryExpression.primary.text == `"Hello, World!"`);
+    });
+
+    // ThrowExpression's parse AssignExpression's to avoid ambiguities in comma-seperated lists
+    testThrow(q{
+        int main()
+        {
+            foo(throw someThrowable, bar);
+        }
+    },
+    (throwExp, throwable)
+    {
+        auto id = throwable.primaryExpression
+                            .identifierOrTemplateInstance
+                            .identifier;
+
+        assert(id.text == "someThrowable");
+    });
+}
```

Comment:
I think you can replace these tests with better readable ones in https://github.com/dlang-community/libdparse/tree/master/test/ast_checks using XML paths and just plain .d files

Reply:
Done